multi_ptr-cast/sycl-2.2/index.md

```diff
@@ -0,0 +1,242 @@
+    (defined through an explicit conversion operator)
+  * Using `static_cast` to cast `multi_ptr<void, Space>` to `multi_ptr<B, Space>`
+    (defined through an explicit conversion operator)
+
+Therefore, the only option to perform a cast from
+`multi_ptr<A, Space>` to `multi_ptr<B, Space>`
+is to use both casts like so:
+```cpp
+// Using global_ptr here to simplify the code
+// decltype(multiPtrA) == global_ptr<A>
+auto multiPtrB = static_cast<global_ptr<B>>(
+                  static_cast<global_ptr<void>>(multiPtrA));
+```
+
+This is problematic on a few levels:
+  * Verbosity
+  * Only allows static casts
+  * Allows casting the underlying `A*` pointer to a `B*` pointer
+    even if the type system forbids it
+  * Does not handle `const` qualifiers
+
+Therefore, there is a clear need to provide more casting options for the `multi_ptr` class
+in order to make the casting safer and easier to use.
+
+## Summary
+
+This proposal folds the `void` specialization of the `multi_ptr` class
+into the main definition of `multi_ptr<ElementType, Space>`,
+handles the cases where `ElementType` is a `const`-qualified type,
+adds a few explicit conversion operators
+as member functions of the `multi_ptr` class,
+and also adds several free functions to the `cl::sycl` namespace
+that follow the naming and semantics of the `std::shared_ptr` pointer cast functions
+defined by the C++17 standard: https://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast.
+
+## Specializations for void and const void
+
+In SYCL 1.2.1, the `void` specialization of the `multi_ptr` class
+is separate from the main definition
+because it cannot contain any reference types.
+For the purpose of the specification,
+we propose folding the two definitions
+into a single definition of the `multi_ptr` class,
+with comments denoting which types are available
+only if `ElementType` is not `void` or `const void`.
+The main reason for this simplification
+is the introduction of the specialization for `const void`,
+which would require the specification to define another specialization
+that would be mostly the same as the specialization for `void`.
+Simplifying the definiton like this
+also covers the case where `ElementType` is not `void` or `const void`,
+but some other `const`-qualified type.
+
+## Explicit conversion operators
+
+The new interface of the `multi_ptr` class would look like this:
+
+```cpp
+namespace cl {
+namespace sycl {
+
+template <typename ElementType, access::address_space Space>
+class multi_ptr {
+ public:
+  /// All existing members here
+
+  ...
+
+  // Explicit conversion to `multi_ptr<ElementTypeU, Space>`
+  template <typename ElementTypeU>
+  explicit operator multi_ptr<ElementTypeU, Space>() const;
+
+  // Implicit conversion to `multi_ptr<const ElementType, Space>`
+  operator multi_ptr<const ElementType, Space>() const;
+};
+
+} // namespace sycl
+} // namespace cl
+```
+
```

Comment:
could be a github effect but formatting seems a bit off

Reply:
Yeah, it does seem strange, I don't know why, VSCode displays it correctly